unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs

```diff
@@ -0,0 +1,191 @@
+    }
+
+    /// <summary>
+    /// Add relevant information on each body part to observations.
+    /// </summary>
+    public void CollectObservationBodyPart(BodyPart bp)
+    {
+        var rb = bp.rb;
+        AddVectorObs(bp.groundContact.touchingGround ? 1 : 0); // Is this bp touching the ground
+        bp.groundContact.touchingGround = false;
+
+        AddVectorObs(rb.velocity);
+        AddVectorObs(rb.angularVelocity);
+        Vector3 localPosRelToHips = hips.InverseTransformPoint(rb.position);
+        AddVectorObs(localPosRelToHips);
+
+        if (bp.joint && (bp.rb.transform != handL && bp.rb.transform != handR))
+        {
+            var jointRotation = GetJointRotation(bp.joint);
+            AddVectorObs(jointRotation); // Get the joint rotation
+        }
+    }
+
+    /// <summary>
+    /// Loop over body parts to add them to observation.
+    /// </summary>
+    public override void CollectObservations()
+    {
+        AddVectorObs(goalDirection);
+        foreach (var bodyPart in bodyParts.Values)
+        {
+            CollectObservationBodyPart(bodyPart);
+        }
+    }
+
+    /// <summary>
+    /// Apply torque to bodyPart `bp` according to defined goal `x, y, z` angle and force `strength`.
+    /// </summary>
+    public void SetNormalizedTargetRotation(BodyPart bp, float x, float y, float z, float strength)
+    {
+        // Transform values from [-1, 1] to [0, 1]
+        x = (x + 1f) * 0.5f;
+        y = (y + 1f) * 0.5f;
+        z = (z + 1f) * 0.5f;
+
+        var xRot = Mathf.Lerp(bp.joint.lowAngularXLimit.limit, bp.joint.highAngularXLimit.limit, x);
+        var yRot = Mathf.Lerp(-bp.joint.angularYLimit.limit, bp.joint.angularYLimit.limit, y);
+        var zRot = Mathf.Lerp(-bp.joint.angularZLimit.limit, bp.joint.angularZLimit.limit, z);
+
+        bp.joint.targetRotation = Quaternion.Euler(xRot, yRot, zRot);
+        var jd = new JointDrive
+        {
+            positionSpring = ((strength + 1f) * 0.5f) * 10000f,
+            maximumForce = 250000f
+        };
+        bp.joint.slerpDrive = jd;
+    }
+
+    public override void AgentAction(float[] vectorAction, string textAction)
+    {
+        // Apply action to all relevant body parts. 
+        SetNormalizedTargetRotation(bodyParts[chest], vectorAction[0], vectorAction[1], vectorAction[2],
+            vectorAction[26]);
+        SetNormalizedTargetRotation(bodyParts[spine], vectorAction[3], vectorAction[4], vectorAction[5],
+            vectorAction[27]);
+
+        SetNormalizedTargetRotation(bodyParts[thighL], vectorAction[6], vectorAction[7], 0, vectorAction[28]);
+        SetNormalizedTargetRotation(bodyParts[shinL], vectorAction[8], 0, 0, vectorAction[29]);
+        SetNormalizedTargetRotation(bodyParts[footL], vectorAction[9], vectorAction[10], vectorAction[11],
+            vectorAction[30]);
+        SetNormalizedTargetRotation(bodyParts[thighR], vectorAction[12], vectorAction[13], 0, vectorAction[31]);
```

Comment:
Why not encapsulate the transform, and the indices into vectorAction inside bodypart? The danger here (apart from having 2 different objects that refer to the same entity from different angles) is that mistakes in this method are easily introduced and can be difficult to catch.

Reply:
So the `BodyPart` class would contain the indices which correspond to `x`, `y`, `z`, and `strength` values, and here instead we do something like `SetNormalizedTargetRotation(bodyPart, vectorAction)`?

Reply:
So move `SetNormalizedTargetRotation` into the body part itself?

Reply:
Yes. Move both the transform and SetNormalizedTargetRotation into BodyPart. Then this method will just enumerate all the body parts and call their SetNormalizedTargetRotation. The problem is that each BodyPart will need to know which indices to use, so the construction of the body parts will be more complicated.
Let's leave it as is for now, I'll try to think of a better way to match indices to body parts.

Reply:
I like this suggestion @vladimiroster. Why would each body part need to be aware of `vectorAction` though?

The `SetNormalizedTargetRotation` would still accept 4 inputs (x,y,z,strength) and it is on the Agent class to retrieve those values from vectorAction/textAction and pass them accordingly.